README.md
Last Resort aims to help your application catch unhandled exceptions and
unhandled rejected promises in the browser.

Platforms Supported
===================

[![Browser Stack](https://www.browserstack.com/images/mail/browserstack-logo-footer.png)](https://www.browserstack.com)

Last Resort is tested using
[BrowserStack](https://www.browserstack.com). BrowserStack provides this service
for free under their program for supporting open-source software.

Last Resort is supported on:

* Desktop: Chrome, Firefox, Edge, IE11, IE10, IE9, Opera, and Safari (on El
Capitan, Yosemite, Mavericks).

* Mobile: iOS 9.3 down to 6. Android 4.4. Presumably later versions of Android
  are also fine, although we've not been able to test them.

We test against the latest versions offered by the vendors of these browsers on
their respective platforms.

Borderline cases:

* The suite runs on IE9 but it does not perform a thorough check of the values
  obtained from the events generated by the browser. There are a few quirks
  there.

If you see a platform you think should be tested but isn't, then chances are
that Browser Stack is not providing support for it, or is not providing stable
support for it. You should inspect the ``karma.conf.js`` file to see platforms
that have been commented out. If you can get these platforms to work, you are
welcome to put in a pull request.

Not supported, and unlikely to be supported, ever:

* Android versions less than 4.4.

* iOS versions less than 6.

**IMPORTANT IMPORTANT IMPORTANT: Last Resort depends on
``<global>.addEventListener("unhandledrejection", ...)`` being supported in your
environment in order to trap unhandled rejections.** If your platform does not
support this, then Last Resort will not be able to trap unhandled
rejections. [This
page](https://developer.mozilla.org/en-US/docs/Web/API/PromiseRejectionEvent)
suggests that only Chrome 49 and over supports it natively. On browsers that do
not support it natively, you must use a promise library that provides support
for ``unhandledrejection``
events. [Bluebird](https://github.com/petkaantonov/bluebird/) is one such
library. It is actually the library used for testing Last Resort's handling of
unhandled rejections.

**Note on using Last Resort in workers**. Last Resort, when used with Bluebird,
is able to trap unhandledrejection in workers. However, it has to use
`self.onunhandledrejection`` rather than ``self.addEventListener``. This is
because a lot of platforms (including **all** Microsoft browsers) do not
actually allow using ``self.addEventListener`` to listen for custom
exceptions. See [this
comment](https://github.com/petkaantonov/bluebird/pull/1213#issuecomment-243168274)
for details.

Loading Last Resort
===================

Last Resort can be loaded in the following ways:

1. In a `script` element. In this case it will be available as `LastResort` on
   the `window` object in which it has been loaded.

2. As an AMD module. It will export the same thing as what `LastResort` contains
   in the scenario above. You may name the module whatever you want, so long as
   you define a path for it in your `paths` configuration.

3. As a CommonJS module. It will export the same as the thing as the earlier
   options.

4. In a `script` element and then as an AMD or CommonJS module. This is not a
   common way to load scripts but Last Resort supports it. When it is loaded by
   `script`, it creates `LastResort` like in the first scenario above. When it
   is then loaded again as a module, it checks whether `LastResort` already
   exists in the global space, and returns *that* if it exists, rather than
   create a new module. If `LastResort` does not exist, then it will initialize
   itself from scratch. Note that, when loaded first through `script` and then
   as a module, `LastResort` will still be present in the global space if it had
   only been loaded through `script`.

Why this 4th method? This is supported in order to allow loading Last Resort as
early as possible, while still allowing it to be referenced as a module, and not
forcing modules that need to use it to refer to a global. Ideally, Last Resort
should be loaded before any other code so that it can report errors as early as
possible. In one project in which it is used, it is loaded with a `script`
element, then we register an absolutely bare error handler that uses `alert` to
report to the user, then more substantial scripts are loaded, including an AMD
module that provides a Bootstrap-based dialog in case of error. This module
loads Last Resort as an AMD module and replaces the initial barebones error
handler with its more sophisticated one.

Using Last Resort
=================

Last Resort listens to these two events:

* ``error``, which is emitted when a thrown exception is not caught by any code.

* ``unhandledrejection``, which is emitted when a rejected promise is not
  handled by any code.

In the following when we say "the events" we are referring to these two events.

Last Resort exports these items:

* ``install(context, options)`` installs Last Resort to intercept the events on
  ``context``. The parameter ``context`` is often ``window`` but could be the
  global context of a worker too (``self``). It returns an instance of
  ``OnError`` this instance is what is now handling the events.

  The ``options`` parameter is a plain object holding possible options:

  + ``force`` when set to ``true`` will force the installation of Last Resort on
     the ``context`` even if it was already installed previously. Otherwise an
     exception is raised. If ``force`` is used and Last Resort was previously
     installed, the previous installation will be automatically uninstalled.

  + ``noUnhandledRejection`` when set to ``true`` will **NOT** install a handler
     to catch unhandled rejections.

* The ``OnError`` class, which has the following methods:

  + ``register(fn)`` registers the function ``fn`` to handle the events. ``fn``
     will be called with the event objects generated for the ``error`` and
     ``unhandledrejection`` events. It should be prepared to examine the event
     to determine what happened exactly. **Until ``register(fn)`` is called, the
     ``OnError`` instance does not do anything when an event is emitted.**

  + ``uninstall()`` uninstalls the ``OnError``

* ``isInstalled(context)`` returns ``true`` if ``install(context)`` was
  ever called. ``false`` otherwise.

* ``wasTriggered()`` returns ``true`` if any event was raised in any context.

By default, the registered function is called both when an uncaught exception
happens and when an unhandled Promise rejection happens. So it has to be ready
to examine its arguments and determine how to process that information depending
on the types and number of the arguments passed.

When the registered function is called, Last Resort calls ``uninstall`` on the
``OnError` object that bears the registered function. This prevents infinite
recursion if it happens that the registered function causes new errors. It is up
to the developer of the registered function to ensure that this function does
not screw things up while it is running. It *could* call ``LastResort.install``
anew if desired and register its own specialized handler. Or it could do its own
error handling.

Some old browsers bypass ``onerror`` handlers attached to ``iframe`` elements
and instead go straight for the one on the root window. It is up to your
application to decide whether it needs to support these old browsers and use
Last Resort appropriately (e.g. detect the issue and install on the root window
instead of installing on an ``iframe``).

Example
=======

Here is a super simple example:

    import * as lr from "last-resort";

    const onerror: lr.OnError = lr.install(window);

    function handle(ev: Event): void {
        // tslint:disable-next-line:no-console
        console.log(ev);
    }

    onerror.register(handle);

This example just dumps the error to the console.

Developing Last Resort
======================

If you produce a pull request run ``gulp test`` first to make sure it is
clean. If you add features, do add tests for them.

<!--
#  LocalWords:  BrowserStack El Capitan iOS addEventListener issuecomment fn
#  LocalWords:  unhandledrejection onunhandledrejection LastResort CommonJS
#  LocalWords:  barebones OnError noUnhandledRejection isInstalled onerror
#  LocalWords:  wasTriggered iframe
-->


CHANGELOG.md
<a name="1.0.1"></a>
## [1.0.1](https://github.com/lddubeau/last-resort/compare/v1.0.0...v1.0.1) (2018-12-03)



<a name="1.0.0"></a>
# 1.0.0 (2018-12-03)


### Bug Fixes

* remove redundant "use strict" ([5946ba2](https://github.com/lddubeau/last-resort/commit/5946ba2))
* upgrade testing framework and revamp it ([83fec14](https://github.com/lddubeau/last-resort/commit/83fec14))





test/test.js
/* global chai beforeEach before afterEach sinon Worker describe it URL Blob
   __karma__ mocha */
/* eslint-disable import/no-unresolved */
import require from "require";
import * as lr from "last-resort";
import Promise from "bluebird";

window.Promise = Promise;

let frame;
let frameWindow;
const { assert } = chai;

const agent = navigator.userAgent;
// const CHROME = agent.indexOf(" Chrome/") !== -1 &&
//         // Edge also contains Chrome/
//         agent.indexOf(" Edge/") === -1;
const IE9 = agent.indexOf("; MSIE 9.0;") !== -1;
const ANDROID_4_0_TO_4_3 = agent.search(/ Android 4.[0-3][^\d]/) !== -1;


if (__karma__.config.slow) {
  // eslint-disable-next-line no-console
  console.log("Increasing timeout!");
  mocha.timeout(5000);
}

beforeEach((done) => {
  frame = document.createElement("iframe");
  frame.id = "test";
  // We tried to load this through a Blob. However, using the Blob prevented the
  // script load events from being generated in loadInFrame.
  document.body.appendChild(frame);
  frameWindow = frame.contentWindow;
  const listener = () => {
    frame.removeEventListener("load", listener);
    done();
  };
  frame.addEventListener("load", listener);
  frame.src = "/base/test/frame.html";
});

afterEach(() => frame.parentNode.removeChild(frame));

function fail() {
  return new Promise((resolve) => {
    // This, on the other hand, we want on the main window. So mocha can trap
    // any errors that occur in the resolution...
    window.setTimeout(resolve, 50);

    frameWindow.trigger();
  });
}


function unhandledReject() {
  frameWindow.reject();
}

// We have to return something because we run with --async-only
const DONE = Promise.resolve();

describe("install", () => {
  it("throws if called twice on the same window without force", () => {
    lr.install(frameWindow);
    assert.throws(
      lr.install.bind(undefined, frameWindow),
      Error,
      /^trying to set onerror more than once on the same context.$/);
    return DONE;
  });

  it("does not throw if called twice on the same window with force", () => {
    lr.install(frameWindow);
    lr.install(frameWindow, { force: true });
    return DONE;
  });

  it("uninstalls the previous handlers if called again with force", () => {
    let onerror = lr.install(frameWindow, { force: true });
    const cb = sinon.spy();
    onerror.register(cb);

    const cb2 = sinon.spy();
    return fail().then(() => {
      assert.isTrue(cb.calledOnce,
                    "the initial callback should have been called once");

      // This removes the old handlers.
      onerror = lr.install(frameWindow, { force: true });
      onerror.register(cb2);
      return fail();
    }).then(() => {
      assert.isTrue(cb.calledOnce,
                    "the initial callback should not have been called again");
      assert.isTrue(cb2.calledOnce,
                    "the second callback should have been called once");
    });
  });
});

function loadInFrame(source) {
  const where = frameWindow.document.getElementsByTagName("head")[0];
  const script = where.ownerDocument.createElement("script");
  script.src = source;
  script.async = true;
  return new Promise((resolve, _reject) => {
    script.addEventListener("load", () => resolve());
    where.appendChild(script);
  });
}

function loadBluebirdInFrame() {
  return loadInFrame(`${require.toUrl("bluebird")}.js`);
}

// We cannot use the last-resort loaded in the main window for some
// tests because its state is not what we want, so we load it into the
// frame and test from that.
function loadLastResortInFrame() {
  return loadInFrame(require.toUrl(`${require.toUrl("last-resort")}.js`))
    .then(loadBluebirdInFrame);
}

function rjsLoadLastResortInFrame() {
  return loadInFrame("/base/node_modules/requirejs/require.js").then(() => {
    frameWindow.require.config({
      baseUrl: "/base",
      paths: {
        bluebird: "node_modules/bluebird/js/browser/bluebird",
        "last-resort": "build/dist/last-resort",
      },
    });
    return new Promise(
      resolve => frameWindow.require(["last-resort", "bluebird"],
                                     (lr_, bb) => {
                                       frameWindow.Promise = bb;
                                       resolve(lr_);
                                     }));
  });
}

describe("wasTriggered", () => {
  it("returns false when there has been no error",
     () => loadLastResortInFrame().then(() => {
       frameWindow.LastResort.install(frameWindow);
       assert.isFalse(frameWindow.LastResort.wasTriggered());
     }));

  it("returns true when there has been an error",
     () => loadLastResortInFrame().then(() => {
       frameWindow.LastResort.install(frameWindow);
       assert.isFalse(frameWindow.LastResort.wasTriggered());

       return fail();
     })
     .then(() => assert.isTrue(frameWindow.LastResort.wasTriggered())));

  it("returns true if there was an unhandled rejection",
     () => loadLastResortInFrame().then(() => {
       const onerror = frameWindow.LastResort.install(frameWindow);
       assert.isFalse(frameWindow.LastResort.wasTriggered());
       return new Promise((resolve) => {
         onerror.register((ev) => {
           ev.preventDefault();
           resolve(ev);
         });
         unhandledReject();
       });
     }).then(() => assert.isTrue(frameWindow.LastResort.wasTriggered())));
});

describe("isInstalled", () => {
  it("returns false when install has not been called on the window", () => {
    assert.isFalse(lr.isInstalled(frameWindow));
    return DONE;
  });

  it("returns true when install has been called on the window", () => {
    lr.install(frameWindow);
    assert.isTrue(lr.isInstalled(frameWindow));
    return DONE;
  });
});

describe("Last Resort", () => {
  it("loads through an AMD loader", () => {
    frameWindow.requirejs = {
      paths: {
        "last-resort": require.toUrl("last-resort"),
      },
    };

    return rjsLoadLastResortInFrame().then(assert.isDefined);
  });

  it("loads through `script` and then through an AMD loader", () => {
    frameWindow.requirejs = {
      paths: {
        "last-resort": require.toUrl("last-resort"),
      },
    };

    return loadLastResortInFrame().then(rjsLoadLastResortInFrame)
      .then((amdLr) => {
        assert.isDefined(amdLr);
        assert.equal(frameWindow.LastResort, amdLr);
      });
  });

  it("causes the registered function to be called", () => {
    const onerror = lr.install(frameWindow);
    const cb = sinon.spy();
    onerror.register(cb);

    return fail().then(() => assert.isTrue(cb.calledOnce));
  });

  it("does not call the registered function more than once when it fails",
     () => {
       const onerror = lr.install(frameWindow);
       let called = 0;
       onerror.register(() => {
         called++;
         // We reinstall so that the error thrown is not causing a Mocha
         // error.
         lr.install(frameWindow);
         frameWindow.trigger();
       });

       return fail().then(() => assert.equal(called, 1));
     });

  it("calls the registered function with the right values on " +
     "uncaught exception", () => {
    const onerror = lr.install(frameWindow);

    return new Promise((resolve) => {
      onerror.register(resolve);
      fail();
    }).then((event) => {
      const { message, filename, lineno, colno } = event;
      const err = event.error;

      //
      // On IE9 we do not check the values. There are a whole bunch of quirky
      // things going on. (e.g. ``message`` is undefined, there's no stack.) If
      // someone wants to update the test to inspect values on IE9, then great.
      //
      // There is a similar problem for Android 4.0 to 4.3.
      //
      if (IE9 || ANDROID_4_0_TO_4_3) {
        return;
      }

      assert.include(["Error: failing on purpose",
                      "Uncaught Error: failing on purpose",
                      "failing on purpose"], message);

      const { errorThrown } = frameWindow;

      // Some plaforms do put stack traces on their Error objects. It is not the
      // role of Last Resort to fix this. So we stop testing here.
      if (!errorThrown.stack) {
        return;
      }

      const match =
              errorThrown.stack.match(/^\s+at (?:.* \()(.*):(\d+):(\d+)\)?$/m) ||
              errorThrown.stack.match(/^.*@(.*):(\d+):(\d+)$/m) ||
              // iOS 6 has no colno, hence the optional colno here.
              errorThrown.stack.match(/^.*@(.*):(\d+)$/m);

      if (!match) {
        throw new Error("cannot match trace");
      }
      let errFilename;
      let errLineno;
      let errColno; // eslint-disable-line no-unused-vars
      if (match) {
        [, errFilename, errLineno, errColno] = match;
      }
      assert.equal(filename, errFilename);
      assert.equal(lineno, errLineno, "the line numbers should match");

      // If there was a colno in the trace, then we must get a colno in the
      // event.
      if (errColno !== undefined) {
        // On Chrome the colno provided to onerror differs from the
        // stack trace. :-/
        assert.isDefined(colno);
      }

      if (err) {
        assert.equal(err.toString(), errorThrown.toString());
      }
    });
  });

  function globalLRTest() {
    const onerror = lr.install(frameWindow);

    return new Promise((resolve) => {
      onerror.register((ev) => {
        ev.preventDefault();
        resolve(ev);
      });
      unhandledReject();
    }).then((ev) => {
      if (frameWindow.PromiseRejectionEvent) {
        assert.isTrue(ev instanceof frameWindow.PromiseRejectionEvent);
      }

      assert.equal(ev.type, "unhandledrejection");

      let { reason, promise } = ev;

      if (!reason && !promise) {
        ({ reason, promise } = ev.detail);
      }

      assert.equal(reason, frameWindow.unhandledRejectError);
      return promise.catch(ex => assert.equal(ex,
                                              frameWindow.unhandledRejectError));
    });
  }

  // On platforms where PromiseRejectionEvent exists, we do not need to load
  // Bluebird to be able to listen to PromiseRejectionEvent. Chrome is one such
  // platform.
  const preExists = (typeof PromiseRejectionEvent !== "undefined");
  it("(global lr) calls the registered function with the right values on " +
     "unhandled rejection",
     preExists ? globalLRTest : () => loadBluebirdInFrame().then(globalLRTest));

  it("does not call the registered function on unhanndled rejection" +
     "if noUnhandledRejection is set", () => {
    const onerror = lr.install(frameWindow, { noUnhandledRejection: true });
    const spy = sinon.spy();
    onerror.register(spy);
    unhandledReject();

    // This is a tricky test. The delay here is arbitrary. The fact is that
    // unhandled rejections are detected asynchronously.
    return Promise.delay(500).then(() => assert.isTrue(spy.notCalled));
  });

  describe("when loaded in a worker causes the registered function", () => {
    let worker;

    before(function before() {
      // This effectively skips the whole `describe`.
      if (typeof Worker === "undefined") {
        this.skip();
      }
    });

    beforeEach(() => {
      worker = new Worker("/base/test/worker.es5.js");
    });

    afterEach(() => worker && worker.terminate());

    // Calling .terminate() is necessary on Firefox for Windows, for some
    // reason...
    it("to be called on uncaught exceptions", () => new Promise((resolve) => {
      worker.onmessage = resolve;
      worker.postMessage("throw");
    }));

    it("to be called on unhandled rejections", () => new Promise((resolve) => {
      worker.onmessage = resolve;
      worker.postMessage("reject");
    }));
  });
});


test/worker.js
self.importScripts("/base/node_modules/bluebird/js/browser/bluebird.js",
                   "/base/build/dist/last-resort.js");

/* global LastResort */
const oe = LastResort.install(self);
oe.register(() => {
  self.postMessage("done");
  self.close();
});

self.onmessage = function onmessage(ev) {
  if (ev.data === "throw") {
    throw new Error("worker failed");
  }

  if (ev.data === "reject") {
    Promise.reject(new Error("rejected"));
  }
};


test/.eslintrc.js
module.exports = {
  extends: "lddubeau-base",
  parserOptions: {
    ecmaVersion: 6,
    sourceType: "module"
  },
  overrides: [
    {
      files: ["worker.js"],
      env: {
        worker: true,
      },
      rules: {
        "no-restricted-globals": "off",
      },
    },
  ]
};


src/last-resort.ts
/**
 * @author Louis-Dominique Dubeau
 * @license MPL 2.0
 * @copyright 2016 Louis-Dominique Dubeau
 */
/* global define module require Promise */
export interface Options {
  noUnhandledRejection: boolean;
  force: boolean;
}

let triggered: boolean = false;

export type Handler = (ev: any) => void;
export type Context = Window;

export class OnError {
  private _context: Context;
  private _registered?: Handler;
  // We keep these fields for diagnosis.
  // @ts-ignore
  private _triggered: boolean = false;
  // @ts-ignore
  private _triggeredUncaughtException: boolean = false;
  // @ts-ignore
  private _triggeredUnhandledException: boolean = false;
  private _onerror: any;
  private _onunhandledrejection: any = null;

  constructor(context: Context, options: Options) {
    this._context = context;
    this._onerror = this._makeErrorHandler();
    this._onunhandledrejection = null;

    context.addEventListener("error", this._onerror);

    if (options.noUnhandledRejection) {
      return;
    }

    const onunhandledrejection: Handler =
      this._onunhandledrejection = this._makeUnhandledRejectionHandler();

    //
    // It is currently impossible to use ``addEventListener`` robustly in
    // workers to listen for ``unhandledrejection``. In brief,
    // ``addEventListener`` works on Chrome, FF and Opera but fails
    // elsewhere. See this for details:
    //
    // https://github.com/petkaantonov/bluebird/pull/1213
    //
    // There could be more sophisticated ways to detect the issue and work
    // around it but I don't want to put more work into this than
    // necessary. This works. We check that we are in a worker. Bluebird has a
    // ``version`` field, a ``getNewLibraryCopy`` function and has a ``Promise``
    // reference on ``Promise``.
    //
    if (!(typeof self !== "undefined" && self.importScripts &&
          Promise.version && Promise.getNewLibraryCopy &&
          Promise.Promise === Promise)) {
      context.addEventListener("unhandledrejection", onunhandledrejection);
    }
    else {
      context.onunhandledrejection = onunhandledrejection;
    }
  }

  public uninstall(): void {
    const context: Context = this._context;

    if (this._onerror) {
      context.removeEventListener("error", this._onerror);
      this._onerror = null;
    }

    if (!this._onunhandledrejection) {
      return;
    }

    context.removeEventListener("unhandledrejection",
                                this._onunhandledrejection);

    // Undo the workaround if necessary.
    if (context.onunhandledrejection === this._onunhandledrejection) {
      context.onunhandledrejection = null;
    }

    this._onunhandledrejection = null;
    delete context.__LastResortInstalledOnError;
  }

  public register(fn: Handler): void {
    this._registered = fn;
  }

  private _handle(evContext: any, ev: any): void {
    const registered = this._registered;
    this._triggered = true;
    triggered = true;

    // Nothing to do!
    if (!registered) {
      return;
    }

    this.uninstall();
    registered.call(evContext, ev);
  }

  private _makeErrorHandler(): Handler {
    return (ev: any) => {
      this._triggeredUncaughtException = true;
      this._handle(this, ev);
    };
  }

  private _makeUnhandledRejectionHandler(): Handler {
    return (ev: any) => {
      this._triggeredUnhandledException = true;
      this._handle(this, ev);
    };
  }
}

export function install(context: Context, options?: Options): OnError {
  options = options || {
    force: false,
    noUnhandledRejection: false,
  };

  const force: boolean = options.force;

  const previous: OnError = context.__LastResortInstalledOnError;
  if (previous) {
    if (!force) {
      throw new Error(
        "trying to set onerror more than once on the same context.");
    }
    previous.uninstall();
  }

  const ret: OnError = context.__LastResortInstalledOnError =
    new OnError(context, options);
  return ret;
}

export function isInstalled(context: Context): boolean {
  return !!context.__LastResortInstalledOnError;
}

export function wasTriggered(): boolean {
  return triggered;
}

export const version: string = "1.0.1";


gulptasks/gulpfile.js
/* global __dirname */
import "babel-polyfill";
import gulp from "gulp";
import { Server } from "karma";
import uglify from "gulp-uglify";
import rename from "gulp-rename";
import del from "del";
import Promise from "bluebird";
import _fs from "fs-extra";
import _child_process from "child_process"; // eslint-disable-line camelcase
import gutil from "gulp-util";
import eslint from "gulp-eslint";
import sourcemaps from "gulp-sourcemaps";
import tslint from "gulp-tslint";
import ts from "gulp-typescript";
import es from "event-stream";
import wrap from "gulp-wrap-js";
import path from "path";

// eslint-disable-next-line camelcase
const child_process = Promise.promisifyAll(_child_process);
const execFileAsync = child_process.execFileAsync;

const fs = Promise.promisifyAll(_fs);
const project = ts.createProject("tsconfig.json");

function exec(command, options) {
  return new Promise((resolve, reject) => {
    child_process.exec(command, options, (err, stdout, stderr) => {
      if (err) {
        gutil.log(stdout);
        gutil.log(stderr);
        reject(err);
      }
      resolve([stdout, stderr]);
    });
  });
}

gulp.task("default", ["dist"]);

const template = fs.readFileSync("src/wrap.js").toString()
      .replace(/\/\*\s*<%= contents %>\s*\*\//, "%= body %");

gulp.task("dist", ["uglify"],
          () => gulp.src("build/src/*").pipe(gulp.dest("dist")));

gulp.task("build", ["tslint"], () => {
  const result = project.src()
          .pipe(sourcemaps.init({ loadMaps: true }))
          .pipe(ts(project));

  return es.merge(result.js.pipe(wrap(template))
                  .pipe(sourcemaps.write("."))
                  .pipe(gulp.dest("build")),
                  result.dts.pipe(gulp.dest("build")));
});

gulp.task("uglify", ["build"], () => gulp.src("build/src/last-resort.js")
          .pipe(sourcemaps.init())
          .pipe(uglify({
            preserveComments: "license",
          }))
          .pipe(rename("last-resort.min.js"))
          // Writing to . prevents inlining the map.
          .pipe(sourcemaps.write("."))
          .pipe(gulp.dest("build/src")));

gulp.task("lint", ["eslint", "tslint"]);

gulp.task("eslint", () =>
          gulp.src(["*.js", "test/**/*.js", "gulptasks/**/*.js"])
          .pipe(eslint())
          .pipe(eslint.format())
          .pipe(eslint.failAfterError()));

gulp.task("tslint", () =>
          gulp.src(["src/**/*.ts"])
          .pipe(tslint())
          .pipe(tslint.report()));

gulp.task("test", ["lint", "versync", "test-karma", "pack"]);

gulp.task("versync",
          () => execFileAsync("./node_modules/.bin/versync", ["-v"]));

gulp.task("test-karma", (done) => {
  new Server({
    configFile: `${__dirname}/../karma.conf.js`,
    singleRun: true,
  }, done).start();
});

gulp.task("pack", ["default"], Promise.coroutine(function *distTask() {
  const dir = "build/pack";
  const t = path.join(dir, "t");
  yield fs.ensureDirAsync(dir);
  yield exec("ln -sf `npm pack ../..` LATEST-DIST.tgz", { cwd: dir });
  // Test that the package can actually be installed.
  yield del(t);
  yield fs.ensureDirAsync(path.join(t, "node_modules"));
  yield exec(`(cd ${t}; npm install ../LATEST-DIST.tgz)`);
  yield del(t);
}));

gulp.task("clean", () => del(["build", "dist"]));


karma.conf.js
/* global module require */
/* eslint-env node */

"use strict";

const thisPackage = require("./package");

// Minimal localConfig if there is not one locally.
let localConfig = {
  browserStack: {},
};
try {
  // eslint-disable-next-line import/no-unresolved, global-require
  localConfig = require("./local_config");
}
catch (ex) {} // eslint-disable-line no-empty

module.exports = function karmaConfig(config) {
  const options = {
    basePath: "",
    frameworks: ["requirejs", "mocha", "chai", "sinon"],
    client: {
      mocha: {
        asyncOnly: true,
        grep: config.grep,
      },
    },
    files: [
      "test/karma-main.js",
      "node_modules/babel-polyfill/dist/polyfill.js",
      { pattern: "build/dist/last-resort.js", included: false },
      { pattern: "test/frame.html", included: false },
      { pattern: "test/!(karma-main).js", included: false },
      {
        pattern: "node_modules/bluebird/js/browser/bluebird.js",
        included: false,
      },
    ],
    exclude: [],
    preprocessors: {
      "test/**/!(karma-main|frame_script|worker).js": ["babelModule"],
      "test/@(karma-main|frame_script|worker).js": ["babel"],
    },
    customPreprocessors: {
      babelModule: {
        base: "babel",
        options: {
          plugins: ["transform-es2015-modules-amd"],
        },
      },
    },
    babelPreprocessor: {
      options: {
        presets: ["es2015"],
        sourceMap: "inline",
      },
      filename: function filename(file) {
        return file.originalPath.replace(/\.js$/, ".es5.js");
      },
      sourceFileName: function sourceFileName(file) {
        return file.originalPath;
      },
    },
    reporters: ["progress"],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ["ChromeHeadless", "FirefoxHeadless"],
    browserStack: {
      project: thisPackage.name,
    },
    customLaunchers: {
      ChromeWin: {
        base: "BrowserStack",
        browser: "Chrome",
        os: "Windows",
        os_version: "10",
      },
      FirefoxWin: {
        base: "BrowserStack",
        browser: "Firefox",
        os: "Windows",
        os_version: "10",
      },
      IE11: {
        base: "BrowserStack",
        browser: "IE",
        browser_version: "11",
        os: "Windows",
        os_version: "10",
      },
      IE10: {
        base: "BrowserStack",
        browser: "IE",
        browser_version: "10",
        os: "Windows",
        os_version: "8",
      },
      IE9: {
        base: "BrowserStack",
        browser: "IE",
        browser_version: "9",
        os: "Windows",
        os_version: "7",
      },
      Edge: {
        base: "BrowserStack",
        browser: "Edge",
        os: "Windows",
        os_version: "10",
      },
      Opera: {
        base: "BrowserStack",
        browser: "Opera",
        os: "Windows",
        os_version: "10",
      },
      SafariElCapitan: {
        base: "BrowserStack",
        browser: "Safari",
        os: "OS X",
        os_version: "El Capitan",
      },
      SafariYosemite: {
        base: "BrowserStack",
        browser: "Safari",
        os: "OS X",
        os_version: "Yosemite",
      },
      SafariMavericks: {
        base: "BrowserStack",
        browser: "Safari",
        os: "OS X",
        os_version: "Mavericks",
      },
      //
      // Tunnel seems to fail on all nexus devices.
      //
      // Android5: {
      //   base: "BrowserStack",
      //   browser: "android",
      //   os: "android",
      //   device: "Google Nexus 6",
      //   os_version: "5.0",
      // },
      //
      Android4_4: {
        base: "BrowserStack",
        browser: "android",
        os: "android",
        device: "Samsung Galaxy Tab 4 10.1",
        os_version: "4.4",
      },
      iOS9_3: {
        base: "BrowserStack",
        browser: "iOS",
        os: "iOS",
        device: "iPhone 6S",
        os_version: "9.3",
      },
      iOS5: {
        base: "BrowserStack",
        browser: "iOS",
        os: "iOS",
        device: "iPhone 5",
        os_version: "6.0",
      },
      //
      // Fails on this platform, in workers. It is unclear why. Could just as
      // well being iSO 5.1 that is unable to pass errors properly to an onerror
      // handler set in a worker.
      //
      // iPhone4S: {
      //   base: "BrowserStack",
      //   browser: "iOS",
      //   os: "iOS",
      //   device: "iPhone 4S",
      //   os_version: "5.1",
      // },
      //
      // Karma cannot connect on this platform...
      //
      // iPhone4: {
      //   base: "BrowserStack",
      //   browser: "iOS",
      //   os: "iOS",
      //   device: "iPhone 4",
      //   os_version: "4.0",
      // },
      //
      // Tunnel seems to fail on this device.
      //
      // WindowsPhone8_1: {
      //   base: "BrowserStack",
      //   browser: "winphone",
      //   os: "winphone",
      //   device: "Nokia Lumia 520",
      //   os_version: "8.1",
      // },
    },
    singleRun: false,
    concurrency: 2,
  };

  // Merge the browserStack configuration we got with the base values in our
  // config.
  Object.assign(options.browserStack, localConfig.browserStack);

  const { browsers } = config;
  if (browsers.length === 1 && browsers[0] === "all") {
    const newList = options.browsers.concat(Object.keys(options.customLaunchers));

    // Yes, we must modify this array in place.
    // eslint-disable-next-line prefer-spread
    browsers.splice.apply(browsers, [0, browsers.length].concat(newList));
  }

  const found = browsers.some((x) => {
    const custom = options.customLaunchers[x];
    return custom && custom.base === "BrowserStack";
  });

  if (found) {
    options.captureTimeout = 3e5;
    options.browserNoActivityTimeout = 3e5;
    options.browserDisconnectTimeout = 3e5;
    options.browserDisconnectTolerance = 3;
    options.client.slow = true;
  }

  config.set(options);
};
